<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic Usage • skater</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic Usage">
<meta property="og:description" content="skater">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">skater</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic_usage.html">Basic Usage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/signaturescience/skater/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic Usage</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/signaturescience/skater/blob/HEAD/vignettes/basic_usage.Rmd" class="external-link"><code>vignettes/basic_usage.Rmd</code></a></small>
      <div class="hidden name"><code>basic_usage.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <code>skater</code> package provides a collection of analysis and
utility functions for <strong>S</strong>NP-based
<strong>k</strong>inship <strong>a</strong>nalysis,
<strong>t</strong>esting, and <strong>e</strong>valuation as an
<strong>R</strong> package. Functions in the package include tools for
working with pedigree data, performing relationship degree inference,
assessing classification accuracy, and summarizing IBD segment data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/signaturescience/skater" class="external-link">skater</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pedigree-parsing-and-manipulation">Pedigree parsing and manipulation<a class="anchor" aria-label="anchor" href="#pedigree-parsing-and-manipulation"></a>
</h2>
<p>Pedigrees define familial relationships in a hierarchical
structure.</p>
<p>One of the formats used by PLINK and other genetic analysis tools is
the <code>.fam</code> file.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> A <code>.fam</code> file is a tabular
format with one row per individual and columns for unique IDs of the
mother, father, and the family unit. The package includes
<code><a href="../reference/read_fam.html">read_fam()</a></code> to read files in this format:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">famfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"3gens.fam"</span>, package<span class="op">=</span><span class="st">"skater"</span>, mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fam</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fam.html">read_fam</a></span><span class="op">(</span><span class="va">famfile</span><span class="op">)</span></span>
<span><span class="va">fam</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 64 × 6</span></span></span>
<span><span class="co">#    fid      id                dadid             momid               sex affected</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1 testped1_g1-b1-s1 0                 0                     1        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1 testped1_g1-b1-i1 0                 0                     2        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1 testped1_g2-b1-s1 0                 0                     1        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1 testped1_g2-b1-i1 testped1_g1-b1-s1 testped1_g1-b1-i1     2        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1 testped1_g2-b2-s1 0                 0                     1        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1 testped1_g2-b2-i1 testped1_g1-b1-s1 testped1_g1-b1-i1     2        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1 testped1_g3-b1-i1 testped1_g2-b1-s1 testped1_g2-b1-i1     2        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1 testped1_g3-b2-i1 testped1_g2-b2-s1 testped1_g2-b2-i1     1        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped2 testped2_g1-b1-s1 0                 0                     2        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped2 testped2_g1-b1-i1 0                 0                     1        1</span></span>
<span><span class="co"># <span style="color: #949494;"># … with 54 more rows</span></span></span></code></pre></div>
<p>Family structures imported from “.fam” formated files can then be
translated to the <code>pedigree</code> structure used by the
<code>kinship2</code> package.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> The “fam” format may include multiple
families, and the <code><a href="../reference/fam2ped.html">fam2ped()</a></code> function will collapse them all
into a <code>tibble</code> with one row per family:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fam2ped.html">fam2ped</a></span><span class="op">(</span><span class="va">fam</span><span class="op">)</span></span>
<span><span class="va">peds</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 8 × 3</span></span></span>
<span><span class="co">#   fid      data             ped       </span></span>
<span><span class="co">#   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span> testped1 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">2</span> testped2 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">3</span> testped3 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">4</span> testped4 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">5</span> testped5 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">6</span> testped6 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">7</span> testped7 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">8</span> testped8 <span style="color: #949494;">&lt;tibble [8 × 5]&gt;</span> <span style="color: #949494;">&lt;pedigree&gt;</span></span></span></code></pre></div>
<p>In the example above, the resulting <code>tibble</code> is nested by
family ID. The <code>data</code> column contains the individual family
information, while the <code>ped</code> column contains the pedigree
object for that family. You can unnest any particular family:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peds</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fid</span><span class="op">==</span><span class="st">"testped1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span>cols<span class="op">=</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 8 × 7</span></span></span>
<span><span class="co">#   fid      id                dadid             momid      sex affec…¹ ped       </span></span>
<span><span class="co">#   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span> testped1 testped1_g1-b1-s1 <span style="color: #BB0000;">NA</span>                <span style="color: #BB0000;">NA</span>           1       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">2</span> testped1 testped1_g1-b1-i1 <span style="color: #BB0000;">NA</span>                <span style="color: #BB0000;">NA</span>           2       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">3</span> testped1 testped1_g2-b1-s1 <span style="color: #BB0000;">NA</span>                <span style="color: #BB0000;">NA</span>           1       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">4</span> testped1 testped1_g2-b1-i1 testped1_g1-b1-s1 testped…     2       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">5</span> testped1 testped1_g2-b2-s1 <span style="color: #BB0000;">NA</span>                <span style="color: #BB0000;">NA</span>           1       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">6</span> testped1 testped1_g2-b2-i1 testped1_g1-b1-s1 testped…     2       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">7</span> testped1 testped1_g3-b1-i1 testped1_g2-b1-s1 testped…     2       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">8</span> testped1 testped1_g3-b2-i1 testped1_g2-b2-s1 testped…     1       1 <span style="color: #949494;">&lt;pedigree&gt;</span></span></span>
<span><span class="co"># <span style="color: #949494;"># … with abbreviated variable name ¹​affected</span></span></span></code></pre></div>
<p>You can also look at a single pedigree:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peds</span><span class="op">$</span><span class="va">ped</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># Pedigree object with 8 subjects</span></span>
<span><span class="co"># Bit size= 4</span></span></code></pre></div>
<p>Or plot that pedigree:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">peds</span><span class="op">$</span><span class="va">ped</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_usage_files/figure-html/plotped-1.png" width="768" style="display: block; margin: auto;"></p>
<p>The <code><a href="../reference/plot_pedigree.html">plot_pedigree()</a></code> function from <code>skater</code>
will iterate over a list of pedigree objects, writing a multi-page PDF,
with each page containing a pedigree from family:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_pedigree.html">plot_pedigree</a></span><span class="op">(</span><span class="va">peds</span><span class="op">$</span><span class="va">ped</span>, file<span class="op">=</span><span class="st">"3gens.ped.pdf"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/ped2kinpair.html">ped2kinpair()</a></code> function takes a pedigree object and
produces a pairwise list of relationships between all individuals in the
data with the expected kinship coefficients for each pair.</p>
<p>The function can be run on a single family:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ped2kinpair.html">ped2kinpair</a></span><span class="op">(</span><span class="va">peds</span><span class="op">$</span><span class="va">ped</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 36 × 3</span></span></span>
<span><span class="co">#    id1               id2                   k</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g1-b1-s1 testped1_g1-b1-s1 0.5  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g1-b1-i1 testped1_g1-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g1-b1-s1 testped1_g2-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1_g1-b1-s1 testped1_g2-b1-i1 0.25 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1_g1-b1-s1 testped1_g2-b2-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g1-b1-s1 testped1_g2-b2-i1 0.25 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g1-b1-s1 testped1_g3-b1-i1 0.125</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g1-b1-s1 testped1_g3-b2-i1 0.125</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1_g1-b1-i1 testped1_g1-b1-i1 0.5  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1_g1-b1-i1 testped1_g2-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #949494;"># … with 26 more rows</span></span></span></code></pre></div>
<p>Or mapped over all families in the pedigree</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kinpairs</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">peds</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pairs<span class="op">=</span><span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">ped</span>, <span class="va">ped2kinpair</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">fid</span>, <span class="va">pairs</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">unnest</a></span><span class="op">(</span>cols<span class="op">=</span><span class="va">pairs</span><span class="op">)</span></span>
<span><span class="va">kinpairs</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 288 × 4</span></span></span>
<span><span class="co">#    fid      id1               id2                   k</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1 testped1_g1-b1-s1 testped1_g1-b1-s1 0.5  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-i1 0.25 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-s1 0    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-i1 0.25 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1 testped1_g1-b1-s1 testped1_g3-b1-i1 0.125</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1 testped1_g1-b1-s1 testped1_g3-b2-i1 0.125</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-i1 0.5  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1 testped1_g1-b1-i1 testped1_g2-b1-s1 0    </span></span>
<span><span class="co"># <span style="color: #949494;"># … with 278 more rows</span></span></span></code></pre></div>
<p>Note that this maps <code><a href="../reference/ped2kinpair.html">ped2kinpair()</a></code> over all
<code>ped</code> objects in the input <code>tibble</code>, and that
relationships are not shown for between-family relationships (which
should all be zero).</p>
</div>
<div class="section level2">
<h2 id="degree-inference">Degree Inference<a class="anchor" aria-label="anchor" href="#degree-inference"></a>
</h2>
<p>The <code>skater</code> package includes functions to translate
kinship coefficients to relationship degrees. The kinship coefficients
could come from <code><a href="../reference/ped2kinpair.html">ped2kinpair()</a></code> or other kinship estimation
software.</p>
<p>The <code><a href="../reference/dibble.html">dibble()</a></code> function creates a <strong>d</strong>egree
<strong>i</strong>nference <code>tibble</code>, with degrees up to the
specified <code>max_degree</code> (default=3), expected kinship
coefficient, and lower (<code>l</code>) and upper (<code>u</code>)
inference ranges as defined in the KING paper.<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> Degree 0 corresponds
to self / identity / monozygotic twins, with an expected kinship
coefficient of 0.5, with inference range &gt;=0.354. Anything beyond the
maximum degree resolution is considered unrelated (degree
<code>NA</code>), with expected kinship coefficient of 0.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dibble.html">dibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 5 × 4</span></span></span>
<span><span class="co">#   degree      k       l      u</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span>      0 0.5     0.354  1     </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">2</span>      1 0.25    0.177  0.354 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">3</span>      2 0.125   0.088<span style="text-decoration: underline;">4</span> 0.177 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">4</span>      3 0.062<span style="text-decoration: underline;">5</span>  0.044<span style="text-decoration: underline;">2</span> 0.088<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">5</span>     <span style="color: #BB0000;">NA</span> 0      -<span style="color: #BB0000;">1</span>      0.044<span style="text-decoration: underline;">2</span></span></span></code></pre></div>
<p>The degree inference <code>max_degree</code> default is 3. Change
this argument to allow more granular degree inference ranges:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dibble.html">dibble</a></span><span class="op">(</span>max_degree <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 7 × 4</span></span></span>
<span><span class="co">#   degree      k       l      u</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span>      0 0.5     0.354  1     </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">2</span>      1 0.25    0.177  0.354 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">3</span>      2 0.125   0.088<span style="text-decoration: underline;">4</span> 0.177 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">4</span>      3 0.062<span style="text-decoration: underline;">5</span>  0.044<span style="text-decoration: underline;">2</span> 0.088<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">5</span>      4 0.031<span style="text-decoration: underline;">2</span>  0.022<span style="text-decoration: underline;">1</span> 0.044<span style="text-decoration: underline;">2</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">6</span>      5 0.015<span style="text-decoration: underline;">6</span>  0.011<span style="text-decoration: underline;">0</span> 0.022<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">7</span>     <span style="color: #BB0000;">NA</span> 0      -<span style="color: #BB0000;">1</span>      0.011<span style="text-decoration: underline;">0</span></span></span></code></pre></div>
<p>Note that the distance between relationship degrees becomes smaller
as the relationship degree becomes more distant. The
<code><a href="../reference/dibble.html">dibble()</a></code> function will throw a warning with
<code>max_degree</code> &gt;=10, and will stop with an error at
&gt;=12.</p>
<p>The <code><a href="../reference/kin2degree.html">kin2degree()</a></code> function infers the relationship degree
given a kinship coefficient and a <code>max_degree</code> up to which
anything more distant is treated as unrelated. Example first degree
relative:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/kin2degree.html">kin2degree</a></span><span class="op">(</span><span class="fl">.25</span>, max_degree<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co"># [1] 1</span></span></code></pre></div>
<p>Example 4th degree relative, but using the default max_degree
resolution of 3:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/kin2degree.html">kin2degree</a></span><span class="op">(</span><span class="fl">.0312</span>, max_degree<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co"># [1] NA</span></span></code></pre></div>
<p>Example 4th degree relative, but increasing the degree
resolution:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/kin2degree.html">kin2degree</a></span><span class="op">(</span><span class="fl">.0312</span>, max_degree<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co"># [1] 4</span></span></code></pre></div>
<p>The <code><a href="../reference/kin2degree.html">kin2degree()</a></code> function is vectorized over values of
<code>k</code>, so it can be used inside of a <code>mutate</code> on a
<code>tibble</code> of kinship coefficients:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get two pairs from each type of relationship we have in kinpairs:</span></span>
<span><span class="va">kinpairs_subset</span> <span class="op">&lt;-</span> </span>
<span>  <span class="va">kinpairs</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">kinpairs_subset</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 10 × 4</span></span></span>
<span><span class="co"># <span style="color: #949494;"># Groups:   k [5]</span></span></span>
<span><span class="co">#    fid      id1               id2                    k</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-s1 0     </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-s1 0     </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1 testped1_g3-b1-i1 testped1_g3-b2-i1 0.062<span style="text-decoration: underline;">5</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped2 testped2_g3-b1-i1 testped2_g3-b2-i1 0.062<span style="text-decoration: underline;">5</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1 testped1_g1-b1-s1 testped1_g3-b1-i1 0.125 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1 testped1_g1-b1-s1 testped1_g3-b2-i1 0.125 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-i1 0.25  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-i1 0.25  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1 testped1_g1-b1-s1 testped1_g1-b1-s1 0.5   </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-i1 0.5</span></span>
<span></span>
<span><span class="co"># Infer degree out to third degree relatives:</span></span>
<span><span class="va">kinpairs_subset</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>degree<span class="op">=</span><span class="fu"><a href="../reference/kin2degree.html">kin2degree</a></span><span class="op">(</span><span class="va">k</span>, max_degree<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 10 × 5</span></span></span>
<span><span class="co"># <span style="color: #949494;"># Groups:   k [5]</span></span></span>
<span><span class="co">#    fid      id1               id2                    k degree</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-s1 0          <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-s1 0          <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1 testped1_g3-b1-i1 testped1_g3-b2-i1 0.062<span style="text-decoration: underline;">5</span>      3</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped2 testped2_g3-b1-i1 testped2_g3-b2-i1 0.062<span style="text-decoration: underline;">5</span>      3</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1 testped1_g1-b1-s1 testped1_g3-b1-i1 0.125       2</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1 testped1_g1-b1-s1 testped1_g3-b2-i1 0.125       2</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-i1 0.25        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-i1 0.25        1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1 testped1_g1-b1-s1 testped1_g1-b1-s1 0.5         0</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-i1 0.5         0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="benchmarking-degree-classification">Benchmarking Degree Classification<a class="anchor" aria-label="anchor" href="#benchmarking-degree-classification"></a>
</h2>
<p>Once estimated kinship is converted to degree, it may be of interest
to compare the inferred degree to truth. When aggregated over many
relationships and inferences, this method can help benchmark performance
of a particular kinship analysis method.</p>
<p>The <code>skater</code> package adapts functionality from the
<code>confusionMatrix</code> package<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> in the <code><a href="../reference/confusion_matrix.html">confusion_matrix()</a></code>
function.</p>
<p>The <code><a href="../reference/confusion_matrix.html">confusion_matrix()</a></code> function on its own outputs a
list with three objects:</p>
<ol style="list-style-type: decimal">
<li>A <code>tibble</code> with calculated accuracy, lower and upper
bounds, the guessing rate and p-value of the accuracy vs. the guessing
rate.</li>
<li>A <code>tibble</code> with the following statistics (for each
class):
<ul>
<li>Sensitivity = A/(A+C)</li>
<li>Specificity = D/(B+D)</li>
<li>Prevalence = (A+C)/(A+B+C+D)</li>
<li>PPV = (sensitivity * prevalence)/((sensitivity * prevalence) +
((1-specificity) * (1-prevalence)))</li>
<li>NPV = (specificity * (1-prevalence))/(((1-sensitivity) * prevalence)
+ ((specificity) * (1-prevalence)))</li>
<li>Detection Rate = A/(A+B+C+D)</li>
<li>Detection Prevalence = (A+B)/(A+B+C+D)</li>
<li>Balanced Accuracy = (sensitivity+specificity)/2</li>
<li>Precision = A/(A+B)</li>
<li>Recall = A/(A+C)</li>
<li>F1 = harmonic mean of precision and recall</li>
<li>False Discovery Rate = 1 - PPV</li>
<li>False Omission Rate = 1 - NPV</li>
<li>False Positive Rate = 1 - Specificity</li>
<li>False Negative Rate = 1 - Sensitivity</li>
</ul>
</li>
<li>A <code>matrix</code> with the contingency table object itself.</li>
<li>A <code>vector</code> with the reciprocal RMSE (R-RMSE). The R-RMSE
is calculated as
<code>sqrt(mean((1/(Target+.5)-1/(Predicted+.5))^2)))</code>, and is a
superior measure to classification accuracy when benchmarking
relationship degree estimation. Taking the reciprocal of the target and
predicted degree results in larger penalties for more egregious
misclassifications (e.g., classifying a first-degree relative pair as
second degree) than misclassifications at more distant relationships
(e.g., misclassifying a fourth-degree relative pair as fifth-degree).
The +0.5 adjustment prevents division-by-zero when a 0th-degree
(identical) relative pair is introduced.</li>
</ol>
<p>To illustrate the usage, first take the <code>kinpairs</code> data
from above and randomly flip ~20% of the true relationship degrees.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function to randomly flip levels of a factor (at 20%, by default)</span></span>
<span><span class="va">randomflip</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">p</span><span class="op">=</span><span class="fl">.2</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">&lt;</span><span class="va">p</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>, <span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Infer degree (truth/target) using kin2degree, then randomly flip 20% of them</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">kinpairs_inferred</span> <span class="op">&lt;-</span> <span class="va">kinpairs</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>degree_truth<span class="op">=</span><span class="fu"><a href="../reference/kin2degree.html">kin2degree</a></span><span class="op">(</span><span class="va">k</span>, max_degree<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>degree_truth<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">degree_truth</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>degree_truth<span class="op">=</span><span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/replace_na.html" class="external-link">replace_na</a></span><span class="op">(</span><span class="va">degree_truth</span>, <span class="st">"unrelated"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>degree_inferred<span class="op">=</span><span class="fu">randomflip</span><span class="op">(</span><span class="va">degree_truth</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">kinpairs_inferred</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 288 × 6</span></span></span>
<span><span class="co">#    fid      id1               id2                   k degree_truth degree_infe…¹</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1 testped1_g1-b1-s1 testped1_g1-b1-s1 0.5   0            0            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-s1 0     unrelated    unrelated    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-s1 0     unrelated    unrelated    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1 testped1_g1-b1-s1 testped1_g2-b1-i1 0.25  1            1            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-s1 0     unrelated    unrelated    </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1 testped1_g1-b1-s1 testped1_g2-b2-i1 0.25  1            1            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1 testped1_g1-b1-s1 testped1_g3-b1-i1 0.125 2            2            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1 testped1_g1-b1-s1 testped1_g3-b2-i1 0.125 2            1            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1 testped1_g1-b1-i1 testped1_g1-b1-i1 0.5   0            0            </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1 testped1_g1-b1-i1 testped1_g2-b1-s1 0     unrelated    unrelated    </span></span>
<span><span class="co"># <span style="color: #949494;"># … with 278 more rows, and abbreviated variable name ¹​degree_inferred</span></span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/confusion_matrix.html">confusion_matrix</a></span><span class="op">(</span>prediction <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_inferred</span>, </span>
<span>                 target <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_truth</span><span class="op">)</span></span>
<span><span class="co"># $Accuracy</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 1 × 5</span></span></span>
<span><span class="co">#   Accuracy `Accuracy LL` `Accuracy UL` `Accuracy Guessing` `Accuracy P-value`</span></span>
<span><span class="co">#      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span>    0.812         0.763         0.856               0.333           1.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-62</span></span></span>
<span><span class="co"># </span></span>
<span><span class="co"># $Other</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 6 × 15</span></span></span>
<span><span class="co">#   Class         N Sensit…¹ Speci…² PPV/P…³   NPV F1/Di…⁴ Preva…⁵ Detec…⁶ Detec…⁷</span></span>
<span><span class="co">#   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">1</span> 0          64      0.75    0.964   0.857 0.931   0.8    0.222   0.167   0.194 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">2</span> 1          72      0.806   0.944   0.829 0.936   0.817  0.25    0.201   0.243 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">3</span> 2          48      0.833   0.967   0.833 0.967   0.833  0.167   0.139   0.167 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">4</span> 3           8      0.75    0.936   0.25  0.992   0.375  0.027<span style="text-decoration: underline;">8</span>  0.020<span style="text-decoration: underline;">8</span>  0.083<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">5</span> unrelated  96      0.854   0.958   0.911 0.929   0.882  0.333   0.285   0.312 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">6</span> Average    57.6    0.799   0.954   0.736 0.951   0.741  0.2     0.162   0.2   </span></span>
<span><span class="co"># <span style="color: #949494;"># … with 5 more variables: `Balanced Accuracy` &lt;dbl&gt;, FDR &lt;dbl&gt;, FOR &lt;dbl&gt;,</span></span></span>
<span><span class="co"># <span style="color: #949494;">#   `FPR/Fallout` &lt;dbl&gt;, FNR &lt;dbl&gt;, and abbreviated variable names</span></span></span>
<span><span class="co"># <span style="color: #949494;">#   ¹​`Sensitivity/Recall/TPR`, ²​`Specificity/TNR`, ³​`PPV/Precision`,</span></span></span>
<span><span class="co"># <span style="color: #949494;">#   ⁴​`F1/Dice`, ⁵​Prevalence, ⁶​`Detection Rate`, ⁷​`Detection Prevalence`</span></span></span>
<span><span class="co"># </span></span>
<span><span class="co"># $Table</span></span>
<span><span class="co">#            Target</span></span>
<span><span class="co"># Predicted    0  1  2  3 unrelated</span></span>
<span><span class="co">#   0         48  4  2  1         1</span></span>
<span><span class="co">#   1          5 58  4  0         3</span></span>
<span><span class="co">#   2          0  3 40  1         4</span></span>
<span><span class="co">#   3          8  4  0  6         6</span></span>
<span><span class="co">#   unrelated  3  3  2  0        82</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># $recip_rmse</span></span>
<span><span class="co"># [1] 0.4665971</span></span></code></pre></div>
<p>You can use <code><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">purrr::pluck()</a></code> to isolate just the
contingency table:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/confusion_matrix.html">confusion_matrix</a></span><span class="op">(</span>prediction <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_inferred</span>, </span>
<span>                 target <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_truth</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"Table"</span><span class="op">)</span></span>
<span><span class="co">#            Target</span></span>
<span><span class="co"># Predicted    0  1  2  3 unrelated</span></span>
<span><span class="co">#   0         48  4  2  1         1</span></span>
<span><span class="co">#   1          5 58  4  0         3</span></span>
<span><span class="co">#   2          0  3 40  1         4</span></span>
<span><span class="co">#   3          8  4  0  6         6</span></span>
<span><span class="co">#   unrelated  3  3  2  0        82</span></span></code></pre></div>
<p>Or optionally output in a tidy (<code>longer=TRUE</code>) format,
then spread stats by class:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/confusion_matrix.html">confusion_matrix</a></span><span class="op">(</span>prediction <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_inferred</span>, </span>
<span>                 target <span class="op">=</span> <span class="va">kinpairs_inferred</span><span class="op">$</span><span class="va">degree_truth</span>, </span>
<span>                 longer <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"Other"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/spread.html" class="external-link">spread</a></span><span class="op">(</span><span class="va">Class</span>, <span class="va">Value</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="va">Average</span>, .after<span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">last_col</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_if</a></span><span class="op">(</span><span class="fu">rlang</span><span class="fu">::</span><span class="va"><a href="https://rlang.r-lib.org/reference/type-predicates.html" class="external-link">is_double</a></span>, <span class="va">signif</span>, <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="33%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="10%">
<col width="14%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left">Statistic</th>
<th align="right">0</th>
<th align="right">1</th>
<th align="right">2</th>
<th align="right">3</th>
<th align="right">unrelated</th>
<th align="right">Average</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Balanced Accuracy</td>
<td align="right">0.860</td>
<td align="right">0.880</td>
<td align="right">0.900</td>
<td align="right">0.8400</td>
<td align="right">0.910</td>
<td align="right">0.880</td>
</tr>
<tr class="even">
<td align="left">Detection Prevalence</td>
<td align="right">0.190</td>
<td align="right">0.240</td>
<td align="right">0.170</td>
<td align="right">0.0830</td>
<td align="right">0.310</td>
<td align="right">0.200</td>
</tr>
<tr class="odd">
<td align="left">Detection Rate</td>
<td align="right">0.170</td>
<td align="right">0.200</td>
<td align="right">0.140</td>
<td align="right">0.0210</td>
<td align="right">0.280</td>
<td align="right">0.160</td>
</tr>
<tr class="even">
<td align="left">F1/Dice</td>
<td align="right">0.800</td>
<td align="right">0.820</td>
<td align="right">0.830</td>
<td align="right">0.3800</td>
<td align="right">0.880</td>
<td align="right">0.740</td>
</tr>
<tr class="odd">
<td align="left">FDR</td>
<td align="right">0.140</td>
<td align="right">0.170</td>
<td align="right">0.170</td>
<td align="right">0.7500</td>
<td align="right">0.089</td>
<td align="right">0.260</td>
</tr>
<tr class="even">
<td align="left">FNR</td>
<td align="right">0.250</td>
<td align="right">0.190</td>
<td align="right">0.170</td>
<td align="right">0.2500</td>
<td align="right">0.150</td>
<td align="right">0.200</td>
</tr>
<tr class="odd">
<td align="left">FOR</td>
<td align="right">0.069</td>
<td align="right">0.064</td>
<td align="right">0.033</td>
<td align="right">0.0076</td>
<td align="right">0.071</td>
<td align="right">0.049</td>
</tr>
<tr class="even">
<td align="left">FPR/Fallout</td>
<td align="right">0.036</td>
<td align="right">0.056</td>
<td align="right">0.033</td>
<td align="right">0.0640</td>
<td align="right">0.042</td>
<td align="right">0.046</td>
</tr>
<tr class="odd">
<td align="left">N</td>
<td align="right">64.000</td>
<td align="right">72.000</td>
<td align="right">48.000</td>
<td align="right">8.0000</td>
<td align="right">96.000</td>
<td align="right">58.000</td>
</tr>
<tr class="even">
<td align="left">NPV</td>
<td align="right">0.930</td>
<td align="right">0.940</td>
<td align="right">0.970</td>
<td align="right">0.9900</td>
<td align="right">0.930</td>
<td align="right">0.950</td>
</tr>
<tr class="odd">
<td align="left">PPV/Precision</td>
<td align="right">0.860</td>
<td align="right">0.830</td>
<td align="right">0.830</td>
<td align="right">0.2500</td>
<td align="right">0.910</td>
<td align="right">0.740</td>
</tr>
<tr class="even">
<td align="left">Prevalence</td>
<td align="right">0.220</td>
<td align="right">0.250</td>
<td align="right">0.170</td>
<td align="right">0.0280</td>
<td align="right">0.330</td>
<td align="right">0.200</td>
</tr>
<tr class="odd">
<td align="left">Sensitivity/Recall/TPR</td>
<td align="right">0.750</td>
<td align="right">0.810</td>
<td align="right">0.830</td>
<td align="right">0.7500</td>
<td align="right">0.850</td>
<td align="right">0.800</td>
</tr>
<tr class="even">
<td align="left">Specificity/TNR</td>
<td align="right">0.960</td>
<td align="right">0.940</td>
<td align="right">0.970</td>
<td align="right">0.9400</td>
<td align="right">0.960</td>
<td align="right">0.950</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="ibd-segment-analysis">IBD Segment Analysis<a class="anchor" aria-label="anchor" href="#ibd-segment-analysis"></a>
</h2>
<p>Tools such as <code>hap-ibd</code><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a> are capable of inferring shared IBD
segments between individuals. The <code>skater</code> package includes
functionality to take those IBD segments, compute shared genomic
centimorgan (cM) length, and convert that shared cM to a kinship
coefficient. In addition to inferred segments, these functions can
estimate “truth” kinship from data simulated by <code>ped-sim</code>.<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a></p>
<p>The <code><a href="../reference/read_ibd.html">read_ibd()</a></code> function reads in the pairwise IBD
segment format. Input to this function can either be inferred IBD
segments from hap-IBD (<code>source="hapibd"</code>) or simulated
segments (<code>source="pedsim"</code>). The first example below uses
data in the <code>hap-ibd</code> output format:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hapibd_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"GBR.sim.ibd.gz"</span>, package<span class="op">=</span><span class="st">"skater"</span>, mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">hapibd_seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_ibd.html">read_ibd</a></span><span class="op">(</span><span class="va">hapibd_fp</span>, source <span class="op">=</span> <span class="st">"hapibd"</span><span class="op">)</span></span>
<span><span class="co"># New names:</span></span>
<span><span class="co"># <span style="color: #00BBBB;">•</span> `1` -&gt; `1...2`</span></span>
<span><span class="co"># <span style="color: #00BBBB;">•</span> `1` -&gt; `1...4`</span></span>
<span><span class="co"># <span style="color: #00BBBB;">•</span> `1` -&gt; `1...5`</span></span>
<span><span class="va">hapibd_seg</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 3,954 × 6</span></span></span>
<span><span class="co">#    id1               id2                 chr     start       end length</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g1-b1-s1 testped1_g3-b1-i1     1 197<span style="text-decoration: underline;">661</span>576 234<span style="text-decoration: underline;">863</span>602   47.1</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g2-b2-i1 testped1_g3-b1-i1     1 197<span style="text-decoration: underline;">661</span>576 231<span style="text-decoration: underline;">017</span>545   39.8</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g3-b1-i1 testped1_g3-b2-i1     1 197<span style="text-decoration: underline;">661</span>576 212<span style="text-decoration: underline;">799</span>139   20.3</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped3_g1-b1-s1 testped3_g3-b2-i1     1   2<span style="text-decoration: underline;">352</span>146  10<span style="text-decoration: underline;">862</span>397   17.7</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped3_g2-b2-i1 testped3_g3-b2-i1     1   2<span style="text-decoration: underline;">352</span>146  10<span style="text-decoration: underline;">862</span>397   17.7</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g1-b1-s1 testped1_g2-b1-i1     1   3<span style="text-decoration: underline;">328</span>659  64<span style="text-decoration: underline;">123</span>868   86.4</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g1-b1-s1 testped1_g3-b1-i1     1   3<span style="text-decoration: underline;">328</span>659  33<span style="text-decoration: underline;">476</span>811   51.2</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g2-b2-s1 testped1_g3-b2-i1     1   5<span style="text-decoration: underline;">003</span>504  32<span style="text-decoration: underline;">315</span>147   45.9</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped2_g1-b1-i1 testped2_g3-b1-i1     1 240<span style="text-decoration: underline;">810</span>528 248<span style="text-decoration: underline;">578</span>622   15.9</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped2_g1-b1-i1 testped2_g2-b2-i1     1 241<span style="text-decoration: underline;">186</span>056 249<span style="text-decoration: underline;">170</span>711   15.5</span></span>
<span><span class="co"># <span style="color: #949494;"># … with 3,944 more rows</span></span></span></code></pre></div>
<p>In order to translate the shared genomic cM length to a kinship
coefficient, you must load a genetic map with <code><a href="../reference/read_map.html">read_map()</a></code>.
Software for IBD segment inference and simulation requires a genetic
map. The map loaded for kinship estimation should be the same one used
for creating the shared IBD segment output. The example below uses a
minimal genetic map created with <code>min_map</code><a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a> that ships with
<code>skater</code>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gmapfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sexspec-avg-min.plink.map"</span>, package<span class="op">=</span><span class="st">"skater"</span>, mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">gmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_map.html">read_map</a></span><span class="op">(</span><span class="va">gmapfile</span><span class="op">)</span></span>
<span><span class="va">gmap</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 28,726 × 3</span></span></span>
<span><span class="co">#      chr  value      bp</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span>     1 0       <span style="text-decoration: underline;">752</span>721</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span>     1 0.029<span style="text-decoration: underline;">2</span> 1<span style="text-decoration: underline;">066</span>029</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span>     1 0.082<span style="text-decoration: underline;">9</span> 1<span style="text-decoration: underline;">099</span>342</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span>     1 0.157  1<span style="text-decoration: underline;">106</span>473</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span>     1 0.246  1<span style="text-decoration: underline;">152</span>631</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span>     1 0.294  1<span style="text-decoration: underline;">314</span>015</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span>     1 0.469  1<span style="text-decoration: underline;">510</span>801</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span>     1 0.991  1<span style="text-decoration: underline;">612</span>540</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span>     1 1.12   1<span style="text-decoration: underline;">892</span>325</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span>     1 1.41   1<span style="text-decoration: underline;">916</span>587</span></span>
<span><span class="co"># <span style="color: #949494;"># … with 28,716 more rows</span></span></span></code></pre></div>
<p>The <code><a href="../reference/ibd2kin.html">ibd2kin()</a></code> function takes the segments and map file
and outputs a <code>tibble</code> with one row per pair of individuals
and columns for individual 1 ID, individual 2 ID, and the kinship
coefficient for the pair:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ibd_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ibd2kin.html">ibd2kin</a></span><span class="op">(</span>.ibd_data<span class="op">=</span><span class="va">hapibd_seg</span>, .map<span class="op">=</span><span class="va">gmap</span><span class="op">)</span></span>
<span><span class="va">ibd_dat</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 196 × 3</span></span></span>
<span><span class="co">#    id1               id2                kinship</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g1-b1-i1 testped1_g1-b1-s1 0.000<span style="text-decoration: underline;">316</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g1-b1-i1 testped1_g2-b1-i1 0.261   </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g1-b1-i1 testped1_g2-b2-i1 0.263   </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1_g1-b1-i1 testped1_g2-b2-s1 0.000<span style="text-decoration: underline;">150</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1_g1-b1-i1 testped1_g3-b1-i1 0.145   </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g1-b1-i1 testped1_g3-b2-i1 0.133   </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g1-b1-i1 testped2_g1-b1-i1 0.000<span style="text-decoration: underline;">165</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g1-b1-i1 testped2_g1-b1-s1 0.000<span style="text-decoration: underline;">323</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1_g1-b1-i1 testped2_g2-b1-i1 0.000<span style="text-decoration: underline;">499</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1_g1-b1-i1 testped2_g2-b1-s1 0.000<span style="text-decoration: underline;">318</span></span></span>
<span><span class="co"># <span style="color: #949494;"># … with 186 more rows</span></span></span></code></pre></div>
<p>As noted above, the IBD segment kinship estimation can be performed
on simulated segments. The package includes an example of IBD data in
that format:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pedsim_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"GBR.sim.seg.gz"</span>, package<span class="op">=</span><span class="st">"skater"</span>, mustWork<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pedsim_seg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_ibd.html">read_ibd</a></span><span class="op">(</span><span class="va">pedsim_fp</span>, source <span class="op">=</span> <span class="st">"pedsim"</span><span class="op">)</span></span>
<span><span class="va">pedsim_seg</span></span>
<span><span class="co"># $IBD1</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 1,553 × 6</span></span></span>
<span><span class="co">#    id1               id2               chr    start       end length</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g1-b1-s1 testped1_g2-b1-i1 1     <span style="text-decoration: underline;">752</span>721 249<span style="text-decoration: underline;">170</span>711   262.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g1-b1-s1 testped1_g2-b1-i1 2     <span style="text-decoration: underline;">118</span>913 243<span style="text-decoration: underline;">043</span>959   249.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g1-b1-s1 testped1_g2-b1-i1 3     <span style="text-decoration: underline;">108</span>226 197<span style="text-decoration: underline;">800</span>244   217.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1_g1-b1-s1 testped1_g2-b1-i1 4     <span style="text-decoration: underline;">167</span>596 190<span style="text-decoration: underline;">936</span>728   200.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1_g1-b1-s1 testped1_g2-b1-i1 5     <span style="text-decoration: underline;">157</span>856 180<span style="text-decoration: underline;">692</span>833   196.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g1-b1-s1 testped1_g2-b1-i1 6     <span style="text-decoration: underline;">183</span>917 170<span style="text-decoration: underline;">981</span>684   184.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g1-b1-s1 testped1_g2-b1-i1 7      <span style="text-decoration: underline;">46</span>239 159<span style="text-decoration: underline;">119</span>486   176.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g1-b1-s1 testped1_g2-b1-i1 8     <span style="text-decoration: underline;">113</span>565 146<span style="text-decoration: underline;">280</span>471   160.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1_g1-b1-s1 testped1_g2-b1-i1 9     <span style="text-decoration: underline;">212</span>908 141<span style="text-decoration: underline;">027</span>939   154.</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1_g1-b1-s1 testped1_g2-b1-i1 10    <span style="text-decoration: underline;">158</span>946 135<span style="text-decoration: underline;">473</span>442   166.</span></span>
<span><span class="co"># <span style="color: #949494;"># … with 1,543 more rows</span></span></span>
<span><span class="co"># </span></span>
<span><span class="co"># $IBD2</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 132 × 6</span></span></span>
<span><span class="co">#    id1               id2               chr       start       end length</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g2-b1-i1 testped1_g2-b2-i1 1     156<span style="text-decoration: underline;">666</span>011 162<span style="text-decoration: underline;">443</span>758  9.43 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g2-b1-i1 testped1_g2-b2-i1 1     197<span style="text-decoration: underline;">638</span>290 213<span style="text-decoration: underline;">685</span>761 20.5  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g2-b1-i1 testped1_g2-b2-i1 1     243<span style="text-decoration: underline;">586</span>697 249<span style="text-decoration: underline;">170</span>711  9.43 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1_g2-b1-i1 testped1_g2-b2-i1 2      40<span style="text-decoration: underline;">779</span>973  67<span style="text-decoration: underline;">697</span>179 25.7  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1_g2-b1-i1 testped1_g2-b2-i1 3      26<span style="text-decoration: underline;">902</span>677  27<span style="text-decoration: underline;">840</span>868  0.797</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g2-b1-i1 testped1_g2-b2-i1 3     186<span style="text-decoration: underline;">680</span>562 192<span style="text-decoration: underline;">093</span>520 12.1  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g2-b1-i1 testped1_g2-b2-i1 4      81<span style="text-decoration: underline;">060</span>970 100<span style="text-decoration: underline;">337</span>853 16.7  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g2-b1-i1 testped1_g2-b2-i1 5      24<span style="text-decoration: underline;">009</span>109  30<span style="text-decoration: underline;">217</span>553  4.83 </span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1_g2-b1-i1 testped1_g2-b2-i1 5      31<span style="text-decoration: underline;">751</span>157 134<span style="text-decoration: underline;">562</span>539 83.7  </span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1_g2-b1-i1 testped1_g2-b2-i1 5     167<span style="text-decoration: underline;">835</span>827 168<span style="text-decoration: underline;">425</span>497  1.15 </span></span>
<span><span class="co"># <span style="color: #949494;"># … with 122 more rows</span></span></span></code></pre></div>
<p>Notably, <code>ped-sim</code> differentiates IBD1 and IBD2 segments.
Given that IBD1 and IBD2 segments are weighted differently in kinship
calculation, this should be accounted for in processing. In the example
below the shared IBD is calculated separately for IBD1 and IBD2 with
<code>type="IBD1"</code> and <code>type="IBD2"</code> respectively. You
can then combine those results and sum the IBD1 and IBD2 kinship
coefficients to get the overall kinship coefficient:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ibd1_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ibd2kin.html">ibd2kin</a></span><span class="op">(</span>.ibd_data<span class="op">=</span><span class="va">pedsim_seg</span><span class="op">$</span><span class="va">IBD1</span>, .map<span class="op">=</span><span class="va">gmap</span>, type<span class="op">=</span><span class="st">"IBD1"</span><span class="op">)</span></span>
<span><span class="va">ibd2_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ibd2kin.html">ibd2kin</a></span><span class="op">(</span>.ibd_data<span class="op">=</span><span class="va">pedsim_seg</span><span class="op">$</span><span class="va">IBD2</span>, .map<span class="op">=</span><span class="va">gmap</span>, type<span class="op">=</span><span class="st">"IBD2"</span><span class="op">)</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">ibd1_dat</span>,<span class="va">ibd2_dat</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">id1</span>,<span class="va">id2</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>kinship <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">kinship</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span></span>
<span><span class="co"># <span style="color: #949494;"># A tibble: 48 × 3</span></span></span>
<span><span class="co">#    id1               id2               kinship</span></span>
<span><span class="co">#    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 1</span> testped1_g1-b1-i1 testped1_g2-b1-i1   0.245</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 2</span> testped1_g1-b1-i1 testped1_g2-b2-i1   0.245</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 3</span> testped1_g1-b1-i1 testped1_g3-b1-i1   0.136</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 4</span> testped1_g1-b1-i1 testped1_g3-b2-i1   0.124</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 5</span> testped1_g1-b1-s1 testped1_g2-b1-i1   0.245</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 6</span> testped1_g1-b1-s1 testped1_g2-b2-i1   0.245</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 7</span> testped1_g1-b1-s1 testped1_g3-b1-i1   0.109</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 8</span> testped1_g1-b1-s1 testped1_g3-b2-i1   0.121</span></span>
<span><span class="co"># <span style="color: #BCBCBC;"> 9</span> testped1_g2-b1-i1 testped1_g2-b2-i1   0.254</span></span>
<span><span class="co"># <span style="color: #BCBCBC;">10</span> testped1_g2-b1-i1 testped1_g3-b1-i1   0.245</span></span>
<span><span class="co"># <span style="color: #949494;"># … with 38 more rows</span></span></span></code></pre></div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="https://www.cog-genomics.org/plink/1.9/formats#fam" class="external-link uri">https://www.cog-genomics.org/plink/1.9/formats#fam</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Sinnwell, Jason P., Terry M. Therneau, and Daniel J.
Schaid. “The kinship2 R package for pedigree data.” <em>Human
heredity</em> 78.2 (2014): 91-93.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Manichaikul, A., Mychaleckyj, J. C., Rich, S. S., Daly,
K., Sale, M., &amp; Chen, W. M. (2010). Robust relationship inference in
genome-wide association studies. Bioinformatics (Oxford, England),
26(22), 2867–2873. <a href="https://doi.org/10.1093/bioinformatics/btq559" class="external-link uri">https://doi.org/10.1093/bioinformatics/btq559</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p><a href="https://github.com/m-clark/confusionMatrix" class="external-link uri">https://github.com/m-clark/confusionMatrix</a><a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p><a href="https://github.com/browning-lab/hap-ibd#output-files" class="external-link uri">https://github.com/browning-lab/hap-ibd#output-files</a><a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p><a href="https://github.com/williamslab/ped-sim#output-ibd-segments-file" class="external-link uri">https://github.com/williamslab/ped-sim#output-ibd-segments-file</a><a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p><a href="https://github.com/williamslab/min_map" class="external-link uri">https://github.com/williamslab/min_map</a><a href="#fnref7" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephen Turner, Matthew Scholz, VP Nagraj.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
